ext/uri/php_uriparser.c

```diff
@@ -0,0 +1,429 @@
+		uriparser_copy_uri(&uriparser_uris->normalized_uri, &uriparser_uris->uri);
+		uriparser_normalize_uri(&uriparser_uris->normalized_uri);
+		uriparser_uris->normalized_uri_initialized = true;
+	}
+}
+
+static UriUriA *uriparser_read_uri(uriparser_uris_t *uriparser_uris, uri_component_read_mode_t read_mode)
+{
+	switch (read_mode) {
+		case URI_COMPONENT_READ_RAW:
+			return &uriparser_uris->uri;
+		case URI_COMPONENT_READ_NORMALIZED_ASCII:
+			ZEND_FALLTHROUGH;
+		case URI_COMPONENT_READ_NORMALIZED_UNICODE:
+			normalize_uri_if_needed(uriparser_uris);
+
+			return &uriparser_uris->normalized_uri;
+		EMPTY_SWITCH_DEFAULT_CASE()
+	}
+}
+
+static zend_result uriparser_read_scheme(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	UriUriA *uriparser_uri = uriparser_read_uri(internal_uri->uri, read_mode);
+	ZEND_ASSERT(uriparser_uri != NULL);
+
+	if (uriparser_uri->scheme.first != NULL && uriparser_uri->scheme.afterLast != NULL) {
+		zend_string *str = zend_string_init(uriparser_uri->scheme.first, get_text_range_length(&uriparser_uri->scheme), false);
+		ZVAL_NEW_STR(retval, str);
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+zend_result uriparser_read_userinfo(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	UriUriA *uriparser_uri = uriparser_read_uri(internal_uri->uri, read_mode);
+	ZEND_ASSERT(uriparser_uri != NULL);
+
+	if (uriparser_uri->userInfo.first != NULL && uriparser_uri->userInfo.afterLast != NULL) {
+		ZVAL_STRINGL(retval, uriparser_uri->userInfo.first, get_text_range_length(&uriparser_uri->userInfo));
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_username(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	UriUriA *uriparser_uri = uriparser_read_uri(internal_uri->uri, read_mode);
+	ZEND_ASSERT(uriparser_uri != NULL);
+
+	if (uriparser_uri->userInfo.first != NULL && uriparser_uri->userInfo.afterLast != NULL) {
+		size_t length = get_text_range_length(&uriparser_uri->userInfo);
+		char *c = memchr(uriparser_uri->userInfo.first, ':', length);
+
+		if (c == NULL && length > 0) {
+			ZVAL_STRINGL(retval, uriparser_uri->userInfo.first, length);
+		} else if (c != NULL && c - uriparser_uri->userInfo.first > 0) {
+			ZVAL_STRINGL(retval, uriparser_uri->userInfo.first, c - uriparser_uri->userInfo.first);
+		} else {
+			ZVAL_NULL(retval);
+		}
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_password(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	UriUriA *uriparser_uri = uriparser_read_uri(internal_uri->uri, read_mode);
+	ZEND_ASSERT(uriparser_uri != NULL);
+
+	if (uriparser_uri->userInfo.first != NULL && uriparser_uri->userInfo.afterLast != NULL) {
+		const char *c = memchr(uriparser_uri->userInfo.first, ':', get_text_range_length(&uriparser_uri->userInfo));
```

Comment:
This implementation is inconsistent with the implementation in `uriparser_read_username`.

Reply:
Are the missing `length` variables the problem? Or in which regard you noticed the inconsistency?

Reply:
The reason why I left out the `length` is because in this case we have two of them: one for the whole component, and one for just the password. The former one is not used anywhere else besides the `memchr` call. The latter one can be extracted to a variable but I'd put it after the said `memchr`, and called it somehow else, so consistency would not be perfect.